pkg/services/store/service.go

```diff
@@ -0,0 +1,130 @@
+package store
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/infra/filestorage"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/registry"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/web"
+)
+
+var grafanaStorageLogger = log.New("grafanaStorageLogger")
+
+const RootPublicStatic = "public-static"
+
+type StorageService interface {
+	registry.BackgroundService
+
+	// HTTP API
+	Browse(c *models.ReqContext) response.Response
+	Upload(c *models.ReqContext) response.Response
+	Delete(c *models.ReqContext) response.Response
+
+	// List folder contents
+	List(ctx context.Context, user *models.SignedInUser, path string) (*data.Frame, error)
+
+	// Called from the UI when a dashboard is saved
+	Read(ctx context.Context, user *models.SignedInUser, path string) (*filestorage.File, error)
```

Comment:
Feels like this needs more strict contracts? Something approaching https://github.com/grafana/grafana/pull/31437/files  ?

Reply:
Any examples with the HTTP api?   These are here so that we can link to them easily from `api.go` -- perhaps we want a layer in there that translates:

```
Browse(c *models.ReqContext) response.Respons
```
to 
```
List(ctx context.Context, user *models.SignedInUser, path string)
```

Fundamentaly, the core API is here https://github.com/grafana/grafana/blob/main/pkg/infra/filestorage/api.go#L149

Reply:
Looks like most things stick their HTTP API stuff directly in the API package:
https://github.com/grafana/grafana/blob/main/pkg/api/preferences.go#L20

I'm not sure what our current best practice is with this, but I agree having the top level HTTP calls in the same service feels wrong :thinking:  maybe I'll take a stab just an HTTP wrapper service that then writes to the storage service

Reply:
I understand, I was maytbe confusing this with rthe interface that we will have many implementations for, which will need to be strictly typed (More so than that FileStorage interface, thinking having more like GetRequest, GetResponse, ListRequest, ListResponse types etc as well, using function signature parameters like "string path" feels fragile, but maybe getting way into too much detail here :) 

Long term the http api should also be typed via something like swagger the unified alerting has done but that can come later.

Reply:
@torkelo / @zserge -- any thoughts on these approaches to make the HTTP API distinct from the service API?
* https://github.com/grafana/grafana/pull/46669
* https://github.com/grafana/grafana/pull/46671

> which will need to be strictly typed 
yes, absolutely -- especially when we are trying to read/write things of known types.  This API is access to raw files.  The current target is images, iconst, and JSON files needed for things like geomap and custom icons

Reply:
I am struggling a bit with how to evaluate this without knowing where it fits in (evolution-wise). Is this the super low-level interface we can build a more typed high-level document/entity interface around or is it just because it's super early better to start simple, or this is just for images & icons and not for anything else? The DataFrame for the list operation seems very bad from a typed contract/ease of use for consumers etc. 

But I assume the details of the contract is something we can refine over the coming months and get into details of that might be better in a design doc

Reply:
> Is this the super low-level interface we can build a more typed high-level document/entity interface around or is it just because it's super early better to start simple, or this is just for images & icons and not for anything else?

this is super low level blob storage.  Immediately used for images/icons/raw json.  We can then add constraints around it to support strongly typed objects (like dashboards), that require validation for write.

Reply:
> But I assume the details of the contract is something we can refine over the coming months and get into details of that might be better in a design doc

totally.  In particular when reading/writing well typed things, it is likely not through this raw API